// marko/dist/runtime/helpers/tags-compat/runtime-dom.js
"use strict";const { _l_ } = require("@internal/components-util");
const {
  S_
} = require("../../components/ComponentsContext");
const defineComponent = require("../../components/defineComponent");
const { r: registerComponent } = require("../../components/registry");
const createRenderer = require("../../components/renderer");
const defaultCreateOut = require("../../createOut");
const morphdom = require("../../vdom/morphdom");
const { _k_ } = require("../../vdom/morphdom/fragment");
const dynamicTag = require("../dynamic-tag");

exports.p = function (domCompat) {
  dynamicTag.bn_ = function tagsToVdom(
  renderer,
  renderBody,
  args,
  global)
  {
    const tagsRenderer = domCompat.resolveRegistered(
      renderer || renderBody,
      global
    );

    if (tagsRenderer && domCompat.isRenderer(tagsRenderer)) {
      return (input, out) => {
        return TagsCompat({ i: args ? args : input, r: tagsRenderer }, out);
      };
    }

    return renderer;
  };

  const TagsCompatId = "tags-compat";
  const TagsCompat = createRenderer(
    function (_, out, componentDef, component) {
      const input = Array.isArray(_.i) ? _.i : [_.i];
      const tagsRenderer = domCompat.resolveRegistered(_.r, global);
      const newNode = domCompat.render(out, component, tagsRenderer, input);

      out.bf(out.aa_, component, !newNode);
      if (newNode) {
        out.node({ bo_: () => newNode });
      }
      out.ef();
    },
    // eslint-disable-next-line no-constant-condition





    {
      t: TagsCompatId
    },
    {}
  );

  registerComponent(TagsCompatId, () => ({
    _: TagsCompat,
    Component: defineComponent(
      {
        onMount: domCompat.runComponentEffects,
        onUpdate: domCompat.runComponentEffects,
        onDestroy: domCompat.runComponentDestroy
      },
      TagsCompat
    )
  }));

  // (
  //   nodeAccessor: Accessor,
  //   dynamicTagAttrs?: IntersectionSignal,
  //   intersection?: IntersectionSignal,
  //   valueWithIntersection?: ValueSignal
  // )

  const rendererCache = new WeakMap();

  domCompat.patchDynamicTag((dynamicTag) => (...args) => {
    const signal = dynamicTag(...args);
    return (scope, renderer, getInput) => {
      return signal(scope, create5to6Renderer(renderer), getInput);
    };
  });

  function create5to6Renderer(renderer) {
    let newRenderer = renderer;
    if (renderer && typeof renderer !== "string") {
      const rendererFromAnywhere =
      renderer._ ||
      renderer.render ||
      renderer.renderer && renderer.renderer.renderer ||
      renderer.renderer;

      if (!domCompat.isRenderer(rendererFromAnywhere || renderer)) {
        newRenderer = rendererCache.get(renderer);
        if (!newRenderer) {
          const { Component } = renderer;
          if (Component) {
            const setCustomEvents = Component.prototype.aD_;
            Component.prototype.aD_ = function (
            customEvents,
            scopeId)
            {
              const global = this.ak_;
              for (const customEvent of customEvents) {
                customEvent[1] = domCompat.resolveRegistered(
                  customEvent[1],
                  global
                );
              }

              setCustomEvents.call(this, customEvents, scopeId);
            };
          }
          newRenderer = domCompat.createRenderer(
            (scope, input) =>
            renderAndMorph(scope, rendererFromAnywhere, renderer, input),
            () => _k_()
          );
          rendererCache.set(renderer, newRenderer);
        }
      }
    }
    return newRenderer;
  }

  domCompat.registerRenderer(create5to6Renderer);
  domCompat.init(require("../serialize-noop").aJ_);

  function renderAndMorph(scope, renderer, renderBody, input) {
    const out = defaultCreateOut();
    let host = domCompat.getStartNode(scope);
    let rootNode = host.fragment;
    if (!rootNode) {
      const component = scope.bp_ =
      _l_[scope.m5c];
      rootNode = component._E_;
      host = rootNode.startNode;
      domCompat.setScopeNodes(host, rootNode.endNode);
    }
    const existingComponent = scope.bp_;
    const componentsContext = S_(out);
    const globalComponentsContext = componentsContext.p_;
    let customEvents;
    let normalizedInput;
    globalComponentsContext.az_ = existingComponent;
    out.sync();
    if (renderer) {
      const [rawInput] = input;
      normalizedInput = {};

      for (const key in rawInput) {
        const value = rawInput[key];
        if (/^on[-A-Z]/.test(key)) {
          if (typeof value === "function") {
            (customEvents || (customEvents = {}))[
            key[2] === "-" ? key.slice(3) : key.slice(2).toLowerCase()] =
            [value];
          }
        } else {
          normalizedInput[key === "content" ? "renderBody" : key] = value;
        }
      }

      renderer(normalizedInput, out);
    } else {
      normalizedInput = input[0];
      RenderBodyComponent({ renderBody, args: input }, out);
    }

    domCompat.queueEffect(scope, () => {
      const targetNode = out.ah_().aA_;
      morphdom(rootNode, targetNode, host, componentsContext);
      const componentDefs = componentsContext.af_(
        getRootNode(host)
      );
      const component = componentDefs[0].s_;
      component._E_ = rootNode;
      component.O_ = normalizedInput;
      component.W_ = customEvents;
      scope.bp_ = component;
    });
  }

  function getRootNode(el) {
    var cur = el;
    while (cur.parentNode) cur = cur.parentNode;
    return cur;
  }

  const RenderBodyComponentId = "renderbody-renderer";
  const RenderBodyComponent = createRenderer(
    function (input, out, _componentDef) {
      dynamicTag(
        out,
        input.renderBody,
        null,
        null,
        input.args,
        null,
        _componentDef,
        "0"
      );
    },
    // eslint-disable-next-line no-constant-condition






    {
      t: RenderBodyComponentId,
      i: true
    },
    {}
  );

  registerComponent(RenderBodyComponentId, () => ({
    _: RenderBodyComponent,
    Component: defineComponent({}, RenderBodyComponent)
  }));
};
